### ISSUE TITLE
Extend Ensemble Copula Coupling to deal with probability input

### ISSUE THREAD

**Issue by gavinevans** [metoppv/improver#30]
As a verifier of post-processing output, I want to use Ensemble Copula Coupling so that output, which is either of a percentile format, or of a probability format (e.g. probability of temperature greater than 10C) is converted into ensemble members, for easier verification.

Related issues: #15 

Make use of Flowerdew, (2014) for background information.

Acceptance criteria:
 * Need to be able to deal with forecasts with a single probability (rather than having a series of probabilities for constructing the Cumulative Distribution Function).
 * Need to linearly interpolate probability values at each threshold to create a Cumulative Distribution Function.
 * Need to be able to have some control on the number of percentiles that will be created. The default number of percentiles would match the number of ensemble members from the raw ensemble.
 * Need to support the number of percentiles and the number of raw ensemble members differing within the reordering step, such that there may be 10 raw ensemble members, but 15 percentiles.
 * Unit tests
 * Include an option for random ordering within the Ensemble reordering step, rather than order according to the raw ensemble. Random ordering should remove spatial corrections.
 * Handling of site-specific forecast data, in addition to gridded forecasts.
 * Deal with percentile to percentile interpolation, for example, in the case of moving from an arbitrary set of percentiles to a regularly-spaced set of percentiles.
 * Provide functionality to rebadge quantised percentiles as ensemble member, if the ensemble reordering step is not desired.

What's out of scope?
 * Refinement of the site-specific forecast processing is dependent upon e.g. #34, #42. A follow-on ticket (#48) will be required to review whether the somewhat speculative spot forecast work done in this ticket is applicable. 
 * Consideration of multi-level bounds for the empirical CDF or the possibility of bounds varying depending upon the domain chosen, or even varying by grid point.

**Comment by fionaRust** [comment]
I have been reviewing this code as a first reviewer, mostly in a technical capacity, although I have some understanding of the science we are trying to implement here and I have tried to ensure the unit tests and code handles the kind of cases we may come across to the best of my understanding.
The pull request I have reviewed is here: https://github.com/metoppv/improver/pull/49
Follow on issues to cover future work are available here. These have been raised as the information required to complete this things fully is not available right now, so they will need to be revisited once all the information is available.
 * https://github.com/metoppv/improver/issues/48 - Reveiw this code again when we know the format of spot data, to ensure this code can cope with the final spot data and that unit tests cover the right things for spot data
 * https://github.com/metoppv/improver/issues/54 - Check that the assumptions for the format of the probability and percentile cubes implemented here match CF conventions and the final format of the IMPROVER output.
 * https://github.com/metoppv/improver/issues/55 - Ensure that ECC can cope with the correct form of probabilities. This will require discussion between Johnathan and Nigel to refine what is required, and a discussion with a developer about how this could be implemented.

**Comment by fionaRust** [comment]
Merged pull request #49 
Some possible future technical debt that needs doing, but not in the short term, as it is more to do with possible performance concerns.

**Comment by fionaRust** [comment]
We encountered problems with codacy as we merged #49. We took several steps to fix this which is outlined below:
1. Firstly #117 was raised to revert Ensemble Copula Coupling extension and merged
2. A new pull request #118  was raised to revert the revert (ie add the original Pull request #49 back in again). 
3. A small unrelated pull request was raised and merged to see if it would retrigger codacy. This was #119.
4. Codacy was back up and working, and raised a few more issues. A new branch including latest master and the branch from #118 was created. Additional corrections were added to this branch and then a PR was raised #123 . This pull request was merged. As #119 had a shared history with #123 it also appeared as merged when #123 was merged.

This ticket is now complete :)
